File: share/example/python/example_script2.py
#! /usr/bin/env python
# coding=utf-8

"""
Run script

Steps:
Step 1: Connect to RPC service, robot login
Step 2: Connect to SCRIPT service, robot login
Step 3: Enter the script file name
Step 4: Open and read the script file
Step 5: Send script content
Step 6: When the planner stops running, stop script execution
"""
import time

import pyaubo_sdk
import os
import sys

robot_ip = "127.0.0.1"  # Server IP address
M_PI = 3.14159265358979323846
robot_rpc_client = pyaubo_sdk.RpcClient()
robot_script_client = pyaubo_sdk.ScriptClient()


if __name__ == '__main__':
    robot_rpc_client.connect(robot_ip, 30004)  # API call: connect to RPC service
    robot_rpc_client.login("aubo", "123456")  # API call: robot login
    robot_script_client.connect(robot_ip, 30002)  # API call: connect to SCRIPT
    robot_script_client.login("aubo", "123456")  # API call: robot login

    # Enter the script file name
    filename = input("Please enter the file name: ")
    if not os.path.exists(filename):
        print("File does not exist!")
        sys.exit(1)

    # Open and read the script file
    file = open(filename)
    str_all = ""
    for line in file:
        str_all += line
    str_all += "\r\n\r\n"
    print("Script content: \n", str_all)
    file.close()

    # Send script content
    robot_script_client.sendString(str_all)

    # When the planner stops running, stop script execution
    while True:
        time.sleep(1)
        runtime_state = robot_rpc_client.getRuntimeMachine().getStatus()
        if runtime_state != pyaubo_sdk.RuntimeState.Running:
            print("Planner status:", runtime_state)
            break
        print("Planner status:", runtime_state)

File: share/example/c++/example_script2.cpp
#include "aubo_sdk/rpc.h"
#include "aubo_sdk/script.h"
#include <fstream>
#include <thread>
#ifdef WIN32
#include <Windows.h>
#endif
using namespace arcs::common_interface;
using namespace arcs::aubo_sdk;
using namespace std;

/**
 * Function: Run local script program
 * Steps:
 * Step 1: Connect to RPC service and log in to the robot arm
 * Step 2: Connect to SCRIPT service and log in to the robot arm
 * Step 3: Enter the script file name or absolute path
 * Case 1: If the lua script is copied to the executable program path (build/bin), enter the script file name
 * Case 2: If the lua script is not in the executable program path (build/bin), enter the absolute path of the script
 * Step 4: Read the file and run the script program. If the file fails to open, exit the program.
 */

#define LOCAL_IP "127.0.0.1"

int main(int argc, char **argv) {
#ifdef WIN32
  // Set Windows console output code page to UTF-8
  SetConsoleOutputCP(CP_UTF8);
#endif
  auto rpc = std::make_shared<RpcClient>();
  // API call: Set RPC timeout
  rpc->setRequestTimeout(1000);
  // API call: Connect to RPC service
  rpc->connect(LOCAL_IP, 30004);
  // API call: Login
  rpc->login("aubo", "123456");

  auto script = std::make_shared<ScriptClient>();
  // API call: Connect to SCRIPT service
  script->connect(LOCAL_IP, 30002);
  // API call: Login
  script->login("aubo", "123456");

  // Enter script file name
  char file_name[20];
  cin >> file_name;

  // Open file
  ifstream file;
  file.open(file_name);
  // If file fails to open, exit program
  if (!file) {
    cout << "open fail." << endl;
    exit(1);
  }

  file.close();

  // API call: Send local script file to controller
  script->sendFile(file_name);

  // Wait for planner to start
  int i = 0;
  while (1) {
    if (i++ > 5) {
      std::cerr << "Planner failed to start" << std::endl;
      return -1;
    }
    if (rpc->getRuntimeMachine()->getStatus() == RuntimeState::Running) {
      std::cout << "Planner started successfully, begin executing script" << std::endl;
      break;
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(5));
  }

  // Wait for planner to stop
  while (1) {
    if (rpc->getRuntimeMachine()->getStatus() == RuntimeState::Stopped) {
      std::cout << "Planner stopped, script execution completed" << std::endl;
      break;
    }
    std::this_thread::sleep_for(std::chrono::milliseconds(5));
  }

  // API call: Logout
  rpc->logout();
  // API call: Disconnect
  rpc->disconnect();

  // API call: Logout
  script->logout();
  // API call: Disconnect
  script->disconnect();

  return 0;
}
